waspc/src/Wasp/Generator/DbGenerator/Jobs.hs

```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
 
   -- NOTE(martin): For this to work on Mac, filepath in the list below must be as it is now - not wrapped in any quotes.
-  let npxPrismaCmd = ["npx", "prisma", "migrate", "dev", "--schema", SP.toFilePath schemaFile]
+  let npxPrismaCmd = ["npx", "--no-install", prismaNpxCommand, "migrate", "dev", "--schema", SP.toFilePath schemaFile]
   let scriptArgs =
         if System.Info.os == "darwin"
           then -- NOTE(martin): On MacOS, command that `script` should execute is treated as multiple arguments.
             ["-Fq", "/dev/null"] ++ npxPrismaCmd
           else -- NOTE(martin): On Linux, command that `script` should execute is treated as one argument.
             ["-feqc", unwords npxPrismaCmd, "/dev/null"]
 
-  runNodeCommandAsJob serverDir "script" scriptArgs J.Db
+  let job = runNodeCommandAsJob serverDir "script" scriptArgs J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) jobOutputChannelForwarder
 
 -- | Runs `prisma studio` - Prisma's db inspector.
 runStudio :: Path' Abs (Dir ProjectRootDir) -> J.Job
 runStudio projectDir = do
   let serverDir = projectDir </> serverRootDirInProjectRootDir
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
 
+  let job =
+        runNodeCommandAsJob
+          serverDir
+          "npx"
+          [ "--no-install",
+            prismaNpxCommand,
+            "studio",
+            "--schema",
+            SP.toFilePath schemaFile
+          ]
+          J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) jobOutputChannelForwarder
+
+-- | Runs `prisma generate` to generate the Prisma client.
+runGenerate :: Path' Abs (Dir ProjectRootDir) -> J.Job
+runGenerate projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+  let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
+
+  let job =
+        runNodeCommandAsJob
+          serverDir
+          "npx"
+          [ "--no-install",
+            prismaNpxCommand,
+            "generate",
+            "--schema",
+            SP.toFilePath schemaFile
+          ]
+          J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) errorOnlyJobOutputChannelForwarder
+
+-- | Runs `npm install` to install dependencies.
+-- May be needed before npx commands if dependencies (e.g. Prisma) are not installed.
+npmInstall :: Path' Abs (Dir ProjectRootDir) -> J.Job
+npmInstall projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+
   runNodeCommandAsJob
     serverDir
-    "npx"
-    [ "prisma",
-      "studio",
-      "--schema",
-      SP.toFilePath schemaFile
-    ]
+    "npm"
+    ["install"]
     J.Db
+
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one more time.
+-- It also takes a function that handles forwarding channel output.
+retryJobOnErrorWith :: J.Job -> J.Job -> ChannelForwarder -> J.Job
```

Comment:
I could think about pulling it higher, but the problem then is each caller would then need to remember to wrap their channel forwarder if they called this wrapper. Also, to one of the other comments in here, the reason we have to be careful with returning the `JobExit` is if we allowed the exit to flow from the job on error, and wanted to retry, the actual client would stop listening. Here is how these are typically used:

```haskell
  (_, dbSetupResult) <- liftIO (concurrently (readJobMessagesAndPrintThemPrefixed chan) (setupServer genProjectDir chan))
```

So we can only send 1 `JobExit` message, and that has to be the last thing we send. If we sent the exit from failed job 1, it would not get anything from retry.

Reply:
In thinking it over, I think as a caller having to filter all but the last exit message may be messy? That was the hope of encapsulating that logic into the function.

Reply:
As a caller of `retryJobOnErrorWith`, it sounds equally complicated to me to pass the channel forwarder to it vs call the channel forwarder on it. But I didn't try coding it so I am not 100% sure.

So let's imagine we don't want to have channel forwarder at all, and we go with default behaviour, which is just forwarding all the output. How would we then want `retryJobOnErrorWith` to work? I guess we would output all the JobOutput, but would make sure to not pass that first JobExit. So basically we forward everything, but silence/ignore the JobExit from first try. The rest of JobExits we forward. Would that be ok default behaviour? We could also in theory silence the output up to that first JobExit.

So what if we had that and then on top of that wanted to do smth different, to silence all output unless there is an error -> how would we do that? Coulnd't we just wrap that channel into another one and ignore all output unless JobExit comes? How is that different than the current ChannelForwarder approach?

THere is one benefit I can see in passing the Channel forwarder to the retryJobOnErrorWIth -> if you would give it extra information, e.g. if you would let it know if it is, at the moment of specific message, in the first try, in the fallback job, or in the retry -> that information might be useful to determine what to show or not to show. For example you could choose to not output the messages from the first try if it fails, but do output the messages from the retry, even if it fails.

Am I missing the point here much, what am I overlooking?

Reply:
At the end it seems to me that it might be beneficial to pass it in like this, if we give it more control. Otherwise I might consider taking it out, just to simplify stuff -> that is what I am trying to achieve with questioning this, to possibly have simpler, individual functions.

Reply:
I tried to clean this up a bit by passing in a forwarding strategy, but keeping the forwarder functions themselves local (to ensure the no JobExit message constraint). I also added a notion of first and second attempt to help the only showing prisma generate output if the command fails on retry. I _think_ this may be about as clean as we can make it, but open for feedback! Thanks